builtin/method_str.py

```diff
@@ -26,100 +26,184 @@
+    pattern.
+    """
+    ere = regex_translate.AsPosixEre(p)
+    if not ere.endswith('$'):
+        ere = ere + '$'
+    return _MatchWithEggexByteRange(s, p, ere, len(s))
 
-class Trim(vm._Callable):
 
-    def __init__(self, trim):
+START = 1
+END = 2
+BOTH = 3
+
+
+class hasAffix(vm._Callable):
+    """ Implements `startsWith()`, `endsWith()`. """
+
+    def __init__(self, anchor):
         # type: (int) -> None
-        self.trim = trim
-
-    def _Left(self, string):
-        # type: (str) -> int
-        i = 0
-        while i < len(string):
-            codepoint = string_ops.DecodeUtf8Char(string, i)
-            if not _IsSpace(codepoint):
-                break
-
-            try:
-                i = string_ops.NextUtf8Char(string, i)
-            except error.Strict:
-                assert False, "DecodeUtf8Char should have caught any encoding errors"
-
-        return i
-
-    def _Right(self, string):
-        # type: (str) -> int
-        i = len(string)
-        while i > 0:
-            # For encoding errors, translate error.Strict to error.Expr to
-            # behave like the other builtin methods.
-            try:
-                prev = string_ops.PreviousUtf8Char(string, i)
-            except error.Strict as e:
-                raise error.Expr(e.msg, e.location)
-
-            codepoint = string_ops.DecodeUtf8Char(string, prev)
-            if not _IsSpace(codepoint):
-                break
-
-            i = prev
-
-        return i
+        assert anchor in (START, END), ("Anchor must be START or END")
+        self.anchor = anchor
 
     def Call(self, rd):
         # type: (typed_args.Reader) -> value_t
+        """
+        string => startsWith(pattern_str)   # => bool
+        string => startsWith(pattern_eggex) # => bool
+        string => endsWith(pattern_str)     # => bool
+        string => endsWith(pattern_eggex)   # => bool
+        """
 
         string = rd.PosStr()
+        pattern_val = rd.PosValue()
+        pattern_str = None  # type: str
+        pattern_eggex = None  # type: value.Eggex
+        with tagswitch(pattern_val) as case:
+            if case(value_e.Eggex):
+                pattern_eggex = cast(value.Eggex, pattern_val)
+            elif case(value_e.Str):
+                pattern_str = cast(value.Str, pattern_val).s
+            else:
+                raise error.TypeErr(pattern_val,
+                                    'expected pattern to be Eggex or Str',
+                                    rd.LeftParenToken())
         rd.Done()
 
-        left = 0
-        right = len(string)
-        if self.trim & TRIM_LEFT:
-            left = self._Left(string)
-        if self.trim & TRIM_RIGHT:
-            right = self._Right(string)
+        matched = False
+        try:
+            if pattern_str is not None:
+                if self.anchor & START:
+                    matched, _, _ = string_ops.StartsWithStrByteRange(
```

Comment:
I had a feeling after you pointed me to doc/unicode.md that you would want bytes and not utf-8 decoding. I guess I just assumed the existing call to Python `startswith` was a placeholder.

(and I thought starting off with string functions would be a good easy intro to the project ... well it is, just also more of an self-education than expected!)

I've switched back to startswith/endswith.

That does leave a question in my mind of what should happen with `trim*()` when it trims bytes in a way that the trimmed string is no longer valid utf-8....

I'll revise the code so that if `x => startsWith(y)` then `x => trimStart(y)` does remove 'y' from the start of 'x'.

(And I've added a spec test to cover that case)

Reply:
It's OK if trim() can leave invalid UTF-8, because you can do say

```
var x = $(echo *.py)
```

and have invalid UTF-8, because filenames are just bytes in Unix

So basically in-memory strings are bytes, and then when we do JSON/J8 serialization, we DECODE UTF-8 to make sure the message is valid